### PR TITLE
Add Support for CentOS 8 Stream

### LINKED ISSUE #2380: [Enhancement] Add Support for CentOS 8 Stream
Currently, watchmaker is failing in the [`yum.py`](https://github.com/plus3it/watchmaker/blob/main/src/watchmaker/workers/yum.py) module's [`get_dist_info` function](https://github.com/plus3it/watchmaker/blob/main/src/watchmaker/workers/yum.py#L55-L105). Presumably, the format of CentOS 8 stream's `/etc/system-release` file doesn't adhere to parsing expectations.

### PR DESCRIPTION
**Pull request by ferricoxide**
Closes #2380 

- Validated against CentOS 8 Stream spel AMI `ami-0a6c4db379b52a51a`
- Validated against RHEL 8 spel (beta) AMI `ami-02e6c5efdce6956d`

**Note:** Due the changes in methodology used to identify target EC2s' distro-name, it will be necessary to update any free-standing `config.yaml` files to include `rhel` in the current `redhat` and `centos` config-map (see 480257e in this PR). The `redhat` map-member is preserved for backwards-compatibility

### CODE REVIEW

**Comment by ferricoxide** [comment]
go codebuild go

**Comment by ferricoxide** [comment]
go codebuild go

File: src/watchmaker/static/config.yaml [hunk 1]
```diff
@@ -34,12 +35,14 @@ linux:
...
         - dist:
             - redhat
+            - rhel
             - centos
           el_version: 7
           url: https://watchmaker.cloudarmor.io/yum.defs/saltstack/salt/3004.2/salt-reposync-el7-python3.repo
         #SaltEL8:
         - dist:
             - redhat
+            - rhel
```
Comment (lorengordon) [inline src/watchmaker/static/config.yaml:45]: Hmm, this doesn't quite map out correctly... this config was matching on `redhat`, but `SUPPORTED_DISTS` was matching `red hat` (with a space). That implies to me that we're missing something about how the config is mapped to the dist. But I haven't been in this area of the code in a long time, so I can't say exactly what we're missing without tracing things more thoroughly.

File: src/watchmaker/workers/yum.py [hunk 2]
```diff
@@ -25,15 +25,15 @@ class Yum(WorkerBase, LinuxPlatformManager):
...
+    SUPPORTED_DISTS = ('amazon', 'centos', 'red hat', 'rhel')
 
     # Pattern used to match against the first line of /etc/system-release. A
     # match will contain two groups: the dist name (e.g. 'red hat' or 'amazon')
     # and the dist version (e.g. '6.8' or '2016.09').
     DIST_PATTERN = re.compile(
         r"^({0})"
         "(?:[^0-9]+)"
-        r"([\d]+[.][\d]+)"
+        r"([\d]|[\d]+[.][\d])"
```
Comment (lorengordon) [inline src/watchmaker/workers/yum.py:36]: With the other changes, `DIST_PATTERN` is not used at all, so can be removed entirely.

File: src/watchmaker/workers/yum.py [hunk 3]
```diff
@@ -54,40 +54,15 @@ def _get_amazon_el_version(self, version):
...
-        # Search the release file for a match against _supported_dists
-        matched = self.DIST_PATTERN.search(release.lower())
-        if matched is None:
-            # Release not supported, exit with error
-            msg = (
-                'Unsupported OS distribution. OS must be one of: {0}'
-                .format(', '.join(self.SUPPORTED_DISTS))
-            )
-            self.log.critical(msg)
-            raise WatchmakerError(msg)
```
Comment (lorengordon) [inline src/watchmaker/workers/yum.py:81]: We should still raise and exit with error if `dist` is not one of the `SUPPORTED_DISTS`.

File: src/watchmaker/workers/yum.py (hunk 2 above)
Comment (ferricoxide) [reply src/watchmaker/workers/yum.py]: Was _thinking_ that, but "wasn't eating anything". 

Removed.

File: src/watchmaker/workers/yum.py (hunk 3 above)
Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:81]: Clapper's whacking my test-EC2s. This _should_ satisfy the requirement:

~~~
       # Error if 'dist' does not found in SUPPORTED_DISTS
       if dist not in SUPPORTED_DISTS
           # Release not supported, exit with error
           msg = (
               'Unsupported OS distribution. OS must be one of: {0}'
               .format(', '.join(self.SUPPORTED_DISTS))
           )
           self.log.critical(msg)
           raise WatchmakerError(msg)
~~~

?

At any rate, once it's after 1830, I'll do test-launches to validate.

**Comment by ferricoxide** [comment]
go codebuild go

File: src/watchmaker/workers/yum.py [hunk 4]
```diff
@@ -54,40 +41,15 @@ def _get_amazon_el_version(self, version):
...
-
-        # Assign dist,version from the match groups tuple, removing any spaces
-        dist, version = (x.replace(' ', '') for x in matched.groups())
```
Comment (lorengordon) [inline src/watchmaker/workers/yum.py:84]: Ahh, lol, _this_ is what was changing `red hat` to `redhat`. Ok, now that we're using `distro`, I think we can just re-map `rhel` to `redhat` explicitly?

File: src/watchmaker/workers/yum.py (hunk 3 above)
Comment (lorengordon) [reply src/watchmaker/workers/yum.py:81]: That looks right, yes

**Comment by ferricoxide** [comment]
go codebuild go

File: src/watchmaker/workers/yum.py (hunk 4 above)
Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:84]: Err.. is that outdated code-content? Doing a search in my editor for `x.replace` gives me:
```
E486: Pattern not found: x.replace
```

File: src/watchmaker/workers/yum.py (hunk 3 above)
Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:81]: Alright, bout to launch new EC2s to see if this makes anything vomit in either RHEL 8 or CentOS 8-stream.

Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:81]: Ok, so `SUPPORTED_DISTS` is defined in the `Yum` class but it's not available to the `get_dist_info` function in that class?

Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:81]: Oh... `self.SUPPORTED_DISTS`?

File: src/watchmaker/workers/yum.py (hunk 4 above)
Comment (lorengordon) [reply src/watchmaker/workers/yum.py:84]: That makes sense, since the comment is on a removed line. ;)

It was intended as providing further details and context on this comment, https://github.com/plus3it/watchmaker/pull/2389/files#r1023345795

File: src/watchmaker/workers/yum.py (hunk 3 above)
Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:81]: Alright, working nao.

**Comment by ferricoxide** [comment]
go codebuild go

File: src/watchmaker/workers/yum.py (hunk 4 above)
Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:84]: Ok, so… Resolve?

File: src/watchmaker/static/config.yaml (hunk 1 above)
Comment (ferricoxide) [reply src/watchmaker/static/config.yaml:45]: /shrug

Comment (ferricoxide) [reply src/watchmaker/static/config.yaml:45]: Within the code, itself, the explicit dist-references are pretty limited:

~~~
$ find . -type f -print0 | xargs -0 grep -E "red(| )hat"
./src/watchmaker/workers/yum.py:    SUPPORTED_DISTS = ('amazon', 'centos', 'red hat', 'rhel')
./src/watchmaker/static/config.yaml:            - redhat
./src/watchmaker/static/config.yaml:            - redhat
./src/watchmaker/static/config.yaml:            - redhat
./docs/configuration.md:        redhat, centos, and amazon.)
./docs/configuration.md:          - redhat
./docs/configuration.md:            - redhat
./docs/configuration.md:            - redhat
~~~

So, wouldn't really know where to look.

File: src/watchmaker/workers/yum.py (hunk 4 above)
Comment (lorengordon) [reply src/watchmaker/workers/yum.py:84]: I think we'll be better off creating a map of `distro.id()` values to watchmaker "config" values. That would mean something like changing `SUPPORTED_DISTS` to be a map:

```
    SUPPORTED_DISTS = {
        'amazon': 'amazon',
        'centos': 'centos',
        'rhel': 'redhat'
    }
```

and then we lookup the dist from that map:

```
        try:
            dist = self.SUPPORTED_DISTS[distro.id()]
        except KeyError:
            # Release not supported, exit with error
            msg = (
                'Unsupported OS distribution. OS must be one of: {0}'
                .format(', '.join(self.SUPPORTED_DISTS.keys()))
            )
            self.log.critical(msg)
            raise WatchmakerError(msg)
```

Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:84]: 👍 

Cool. I'll give that a shot – possibly as early as later this afternoon or early this evening.

File: src/watchmaker/static/config.yaml (hunk 1 above)
Comment (ferricoxide) [reply src/watchmaker/static/config.yaml:45]: Ah. When I'd clicked on the link, the page that rendered was hard to follow. May have been a rendering-blip. 🤷

File: src/watchmaker/workers/yum.py (hunk 4 above)
Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:84]: Ok, so, we're initially defining `dist` as the string returned by `distro.id()` (at line #72), but subsequently redefining it with the `try`/`except` block. Is that what we actually want to do? Or do we want to replace the current:

~~~
    def get_dist_info(self):
        """Validate the Linux distro and return info about the distribution."""
        dist = distro.id()
        version = distro.version()[0]
        el_version = None

        # Determine el_version
        if dist == 'amazon':
            el_version = self._get_amazon_el_version(version)
        else:
            el_version = distro.version()[0]

        if el_version is None:
            msg = (
                'Unsupported OS version! dist = {0}, version = {1}.'
                .format(dist, version)
            )
            self.log.critical(msg)
            raise WatchmakerError(msg)

        # Error if 'dist' is not found in SUPPORTED_DISTS
        if dist not in self.SUPPORTED_DISTS:
            # Release not supported, exit with error
            msg = (
                'Unsupported OS distribution. OS must be one of: {0}'
                .format(', '.join(self.SUPPORTED_DISTS))
            )
            self.log.critical(msg)
            raise WatchmakerError(msg)

        dist_info = {
            'dist': dist,
            'el_version': el_version
        }
        self.log.debug('dist_info=%s', dist_info)
        return dist_info
~~~

With something more like:

~~~
    def get_dist_info(self):
        """Validate the Linux distro and return info about the distribution."""
        # Error if 'dist' is not found in SUPPORTED_DISTS
        try:
            dist = self.SUPPORTED_DISTS[distro.id()]
        except KeyError:
            # Release not supported, exit with error
            msg = (
                'Unsupported OS distribution. OS must be one of: {0}'
                .format(', '.join(self.SUPPORTED_DISTS.keys()))
            )
            self.log.critical(msg)
            raise WatchmakerError(msg)
        version = distro.version()[0]
        el_version = None

        # Determine el_version
        if dist == 'amazon':
            el_version = self._get_amazon_el_version(version)
        else:
            el_version = distro.version()[0]

        if el_version is None:
            msg = (
                'Unsupported OS version! dist = {0}, version = {1}.'
                .format(dist, version)
            )
            self.log.critical(msg)
            raise WatchmakerError(msg)

        dist_info = {
            'dist': dist,
            'el_version': el_version
        }
        self.log.debug('dist_info=%s', dist_info)
        return dist_info
~~~

I'm not generally a fan of overloading variables, so, I'll implement the latter and see what breaks.

Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:84]: Alright. The latter works.

**Comment by ferricoxide** [comment]
go codebuild go

**Comment by ferricoxide** [comment]
go codebuild go

File: src/watchmaker/static/config.yaml (hunk 1 above)
Comment (ferricoxide) [reply src/watchmaker/static/config.yaml:45]: Alright, ea9cf8c appears to address this (tested with a `config.yaml` that omits the `rhel` key)

Comment (ferricoxide) [reply src/watchmaker/static/config.yaml:45]: Oh... need to undo this change. 😛

**Comment by ferricoxide** [comment]
go codebuild go

File: src/watchmaker/workers/yum.py (hunk 4 above)
Comment (lorengordon) [reply src/watchmaker/workers/yum.py:84]: Indeed, my suggestion was targeted at replacing the initial assignment of `dist = distro.id()`.

File: src/watchmaker/workers/yum.py [hunk 5]
```diff
@@ -54,40 +45,25 @@ def _get_amazon_el_version(self, version):
 
     def get_dist_info(self):
         """Validate the Linux distro and return info about the distribution."""
-        dist = None
```
Comment (lorengordon) [inline src/watchmaker/workers/yum.py:57]: These three assignments for `dist`, `version`, and `el_version` should probably be kept here at the top, to ensure they are set and have values (`None`). Otherwise, they are set in `try` or `if` blocks, which means they may not be set at all if there is an error or an unexpected condition.

File: src/watchmaker/workers/yum.py [hunk 6]
```diff
@@ -54,40 +45,25 @@ def _get_amazon_el_version(self, version):
...
-                .format(', '.join(self.SUPPORTED_DISTS))
+                .format(', '.join(self.SUPPORTED_DISTS.keys()))
             )
             self.log.critical(msg)
             raise WatchmakerError(msg)
-
-        # Assign dist,version from the match groups tuple, removing any spaces
-        dist, version = (x.replace(' ', '') for x in matched.groups())
+        version = distro.version()[0]
+        el_version = None
```
Comment (lorengordon) [inline src/watchmaker/workers/yum.py:60]: Once this assignment is again handled at the top of the function, this line can be removed.

File: src/watchmaker/workers/yum.py (hunk 5 above)
Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:57]: The `except` should abort the overall execution, though. (and the `version` and `el_version` are still kept at the top, just that `dist` is now a multi-line value-setting.

File: src/watchmaker/workers/yum.py (hunk 6 above)
Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:60]: "This line"?

Comment (lorengordon) [reply src/watchmaker/workers/yum.py:60]: Comments code code in github pull requests are on lines. So yes, "this line" is the line this comment is attached to. It's difficult to refer to specific line numbers, since as changes are made, the line number will change. But at the moment, it is line 60.

File: src/watchmaker/workers/yum.py [hunk 7]
```diff
@@ -104,6 +70,20 @@ def get_dist_info(self):
         self.log.debug('dist_info=%s', dist_info)
         return dist_info
 
+    def get_mapped_dist_name(self):
+        """Return a normalized dist-name value."""
+        # Error if 'dist' is not found in SUPPORTED_DISTS
+        try:
+            dist = self.SUPPORTED_DISTS[distro.id()]
```
Comment (lorengordon) [inline src/watchmaker/workers/yum.py:77]:
```suggestion
            return self.SUPPORTED_DISTS[distro.id()]
```

Comment (ferricoxide) [reply src/watchmaker/workers/yum.py:77]: DAMMIT. 😛

File: src/watchmaker/workers/yum.py (hunk 5 above)
Comment (lorengordon) [reply src/watchmaker/workers/yum.py:57]: The change to use a function simplifies it enough for me. Just need to fix it so it returns the value...

File: src/watchmaker/workers/yum.py (hunk 6 above)
Comment (lorengordon) [reply src/watchmaker/workers/yum.py:50]: Become moot, since the refactor to use a function.

**Comment by ferricoxide** [comment]
go codebuild go

**Comment by lorengordon** [comment]
go codebuild go

**Comment by lorengordon** [comment]
go codebuild go

**Review by lorengordon — APPROVED** [review]
🚀
